guardian/aws-canaries
# Trigger CloudFormation to run whenever Lambda code is changed

## What is the problem?

In the src folder of the repository are three lambda functions, each designed to be run in one or more regions as AWS Synthetics Canaries. When we make a change to one or more of these lambda files, then create a branch and push, we would expect the test canaries in AWS to be updated with our new code. Similarly, when we merge code to main, we would expect our production canaries to be updated. However, there is a problem. When CloudFormation receives instructions to update the canary stack, it will not update the canary if the changeset is empty. The lambda code is stored in a bucket in S3 and referenced in the yaml template. This location in S3 doesn’t change, since we overwrite old versions, so CloudFormation doesn’t detect any changes to the stack and therefore doesn’t run the update.

## What is the solution?

We need a way to update the CloudFormation template whenever there is a change to any of the lambda files, so that there is a non-zero changeset when CloudFormation runs the templates. We can do this by generating the SHA of each of the files that contain canary lambda code, then generating the SHA's of that to get one SHA that represents all lambda code files. Then, we pass this SHA as a parameter to the CDK template in the form of a tag. Now, when we make a change to any of these lambda files, the SHA will be updated, CloudFormation will spot that a tags value has changed and will rerun the canary.

## How to test

Make a change to one of the files in the /src folder, create a new branch and push your changes. The CODE canary should be updated with your change.
Make a change that does not affect files in the /src folder or the generated template, i.e. add a comment to the CDK file or update the README. Push your change. The CODE canary should NOT be updated. You can check this in Riffraff or CloudFront.
